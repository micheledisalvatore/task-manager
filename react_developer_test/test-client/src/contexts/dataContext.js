import React, {
  createContext, useEffect, useMemo, useState,
} from 'react';
import PropType from 'prop-types';
import useFetch from '../hooks/useFetch';

const dataContext = createContext({});

export default dataContext;

export const DataProvider = ({ children }) => {
  const [userName, setUserName] = useState('');
  const [error, setError] = useState(null);
  const { data: usersData = [], loading: usersLoading, error: usersError } = useFetch('https://jsonplaceholder.typicode.com/users');
  const { data: todosData, loading: todosLoading, error: todosError } = useFetch('https://jsonplaceholder.typicode.com/todos');
  const loading = useMemo(() => usersLoading || todosLoading, [usersLoading, todosLoading]);

  useEffect(() => setError(todosError || usersError), [todosError, usersError]);

  const user = useMemo(() => {
    if (usersData && userName.length) {
      const foundUser = usersData.find(({ username }) => username === userName);

      if (foundUser) {
        return foundUser;
      }

      setError('User not found');
    }

    return {};
  }, [usersData, userName]);

  const tasks = useMemo(() => {
    if (todosData && user) {
      const { id: userId } = user;

      const foundTodos = todosData.filter((task) => task.userId === userId);

      if (foundTodos) {
        return foundTodos;
      }

      setError('Todos not found');
    }

    return [];
  }, [user, todosData]);

  const value = {
    onUserNameChange: setUserName,
    user,
    tasks,
    loading,
    error,
  };

  return <dataContext.Provider value={value}>{children}</dataContext.Provider>;
};

DataProvider.propTypes = {
  children: PropType.node.isRequired,
};
