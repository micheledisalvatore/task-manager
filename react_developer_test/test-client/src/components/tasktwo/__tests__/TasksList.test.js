import React from 'react';
import { render, screen } from '@testing-library/react';
import TasksListItem from '../TasksList';
import dataContext from '../../../contexts/dataContext';
import '@testing-library/jest-dom';

jest.mock('../Tick', () => jest.fn(({ className }) => (
  <div
    data-testid="tick"
    className={className}
  />
)));

test('renders the basic component', () => {
  const { container } = render(
    <dataContext.Provider
      value={{
        tasks: [],
      }}
    >
      <TasksListItem />
    </dataContext.Provider>
  );
  expect(container).toMatchSnapshot();

  expect(screen.queryByTestId('list')).toBeEmpty();
});

test('renders the component with 2 tasks', () => {
  const { container } = render(
    <dataContext.Provider
      value={{
        tasks: [
          { id: 0, title: 'Foo title', completed: false },
          { id: 1, title: 'Bar title', completed: true },
        ],
      }}
    >
      <TasksListItem />
    </dataContext.Provider>
  );

  expect(container).toMatchSnapshot();

  expect(screen.queryByTestId('list')).not.toBeEmpty();

  const items = screen.getAllByTestId('item');
  expect(items[0]).toBeInTheDocument();

  const ticks = screen.getAllByTestId('tick');
  expect(ticks[0]).not.toHaveClass('TasksList__Item__Tick--visible');
  expect(ticks[1]).toHaveClass('TasksList__Item__Tick--visible');
});
