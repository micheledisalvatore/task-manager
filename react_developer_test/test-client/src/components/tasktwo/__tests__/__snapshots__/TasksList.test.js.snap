// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the basic component 1`] = `
<div>
  <ul
    class="TasksList"
    data-testid="list"
  />
</div>
`;

exports[`renders the component with 2 tasks 1`] = `
<div>
  <ul
    class="TasksList"
    data-testid="list"
  >
    <li
      class="TasksList__Item"
      data-testid="item"
    >
      <div
        class="TasksList__Item__Tick "
        data-testid="tick"
      />
      Foo title
    </li>
    <li
      class="TasksList__Item"
      data-testid="item"
    >
      <div
        class="TasksList__Item__Tick TasksList__Item__Tick--visible"
        data-testid="tick"
      />
      Bar title
    </li>
  </ul>
</div>
`;
